src/metrics.rs

```diff
@@ -0,0 +1,250 @@
+                        info!("metrics_thread: flush");
+                        if !points.is_empty() {
+                            writer.write(&mut points);
+                            last_write_time = Instant::now();
+                            assert!(points.is_empty());
+                        }
+                        barrier.wait();
+                    }
+                    MetricsCommand::Submit(point) => {
+                        info!("metrics_thread: submit {:?}", point);
+                        points.push(point);
+                    }
+                },
+                Err(RecvTimeoutError::Timeout) => {
+                    trace!("metrics_thread: receive timeout");
+                }
+                Err(RecvTimeoutError::Disconnected) => {
+                    info!("metrics_thread: sender disconnected");
+                    break;
+                }
+            }
+
+            let now = Instant::now();
+            if now.duration_since(last_write_time) >= write_frequency {
+                if !points.is_empty() {
+                    info!("metrics_thread: writing {} points", points.len());
+                    writer.write(&mut points);
+                    last_write_time = now;
+                    assert!(points.is_empty());
+                }
+            }
+        }
+        debug!("metrics_thread: exit");
+    }
+
+    pub fn submit(&self, point: Point) {
+        info!("submitting point: {}", point);
+        self.sender.send(MetricsCommand::Submit(point)).unwrap();
+    }
+
+    pub fn flush(&self) {
+        let barrier = Arc::new(Barrier::new(2));
+        self.sender
+            .send(MetricsCommand::Flush(Arc::clone(&barrier)))
+            .unwrap();
+
+        barrier.wait();
+    }
+}
+
+impl Drop for MetricsAgent {
+    fn drop(&mut self) {
+        self.flush();
+    }
+}
+
+fn get_singleton_agent() -> Arc<Mutex<MetricsAgent>> {
+    static INIT: Once = ONCE_INIT;
+    static mut AGENT: Option<Arc<Mutex<MetricsAgent>>> = None;
+    unsafe {
+        INIT.call_once(|| AGENT = Some(Arc::new(Mutex::new(MetricsAgent::new()))));
+        match AGENT {
+            Some(ref agent) => Arc::clone(&agent),
+            None => panic!("Failed to initialize metrics agent"),
+        }
+    }
+}
+
+/// Submits a new point from any thread.  Note that points are internally queued
+/// and transmitted periodically in batches
+pub fn submit(point: Point) {
+    let agent_mutex = get_singleton_agent();
+    let agent = agent_mutex.lock().unwrap();
+    agent.submit(point);
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use std::sync::atomic::{AtomicUsize, Ordering};
+
+    struct MockMetricsWriter {
+        points_written: AtomicUsize,
+    }
+    impl MockMetricsWriter {
+        fn new() -> MockMetricsWriter {
+            MockMetricsWriter {
+                points_written: AtomicUsize::new(0),
+            }
+        }
+
+        fn assert_points_written(&self, n: usize) {
```

Comment:
Can you make this `fn points_written(&self) -> usize` instead? Just a style thing, but I have a strong preference for tests written as `assert_eq(do_stuff(), expected_value)`. I find it closer to property-based testing, as well as lets me grok the intent of the test without grokking `do_stuff()`.

Reply:
✔️